### PR TITLE
add tests to ensure the sdk is well-formed

### PR DESCRIPTION
**Pull request by devoncarew**
This adds tests to ensure that the Flutter Dart SDK we analyze against is well-formed (has the expected libraries).

This is a follow up to https://github.com/flutter/engine/pull/12435, and part of the work for https://github.com/flutter/flutter/issues/35588.

@jonahwilliams @zanderso

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/dart/sdk_validation_test.dart [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    Directory tempDir;
+
+    setUpAll(() {
+      Cache.disableLocking();
+      tempDir =
+          fs.systemTempDirectory.createTempSync('sdk_validation_test').absolute;
+    });
+
+    tearDownAll(() {
+      tryToDelete(tempDir);
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/dart/sdk_validation_test.dart]: Re-enable cache locking here.

File: packages/flutter_tools/test/general.shard/dart/sdk_validation_test.dart [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  final File dartFile =
+      fs.file(fs.path.join(directory.path, 'lib', 'main.dart'));
+  dartFile.parent.createSync();
+  dartFile.writeAsStringSync(dartSource);
+}
+
+Future<int> analyze(AnalysisServer server, {bool printErrors = true}) async {
+  int errorCount = 0;
+  final Future<bool> onDone =
+      server.onAnalyzing.where((bool analyzing) => analyzing == false).first;
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/dart/sdk_validation_test.dart]: I think we were discussing on another issue that this was not exactly the right way to detect that the analysis server is finished reporting results. If the tool's `AnalysisServer` doesn't have a more correct way to detect the end of the error stream from the analysis server, could these tests become flaky?

File: packages/flutter_tools/test/general.shard/dart/sdk_validation_test.dart (hunk 1 above)
Comment (devoncarew) [reply packages/flutter_tools/test/general.shard/dart/sdk_validation_test.dart]: done

File: packages/flutter_tools/test/general.shard/dart/sdk_validation_test.dart (hunk 2 above)
Comment (devoncarew) [reply packages/flutter_tools/test/general.shard/dart/sdk_validation_test.dart]: It's not part of the analysis server protocol spec, but it does work for all cases modulo when you're using analysis plugins. We do rely on this in other places in the source, including `lib/src/commands/analyze_continuously.dart` and analyze_once.dart.

**Review by zanderso — APPROVED** [review]

**Comment by devoncarew** [comment]
I don't think the code coverage change is legitimate - this PR only adds tests, so shouldn't cause code coverage to go down in unrelated flutter_tools/lib/ files.

File: packages/flutter_tools/test/general.shard/dart/sdk_validation_test.dart [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+void main() {
+  group('sdk validation', () {
+    AnalysisServer server;
+    Directory tempDir;
+
+    setUpAll(() {
+      Cache.disableLocking();
+      tempDir =
+          fs.systemTempDirectory.createTempSync('sdk_validation_test').absolute;
```
Comment (Hixie) [inline packages/flutter_tools/test/general.shard/dart/sdk_validation_test.dart]: @devoncarew @zanderso 
nits: please see these sections of the style guide
https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#prefer-a-maximum-line-length-of-80-characters
https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#have-good-hygiene-when-using-temporary-directories

**Pull request merged**